# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two)

<function add_two at 0x00000213C1DBA340>


In [2]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [4]:
print(lambda x: x+2)

<function <lambda> at 0x00000213C7427880>


In [5]:
add_two(10)

12

In [7]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)

(lambda x: x + 2)(10)

12

In [8]:
num = 25

num_plus_two = (lambda x: x + 2)(num)

print(num_plus_two)

27


#### Saving to a Variable

In [9]:
named_lambda = lambda z: z ** 2

print(named_lambda)

<function <lambda> at 0x00000213C7426E80>


In [11]:
named_lambda(5)

25

#### Multiple Inputs

In [13]:
# Syntax: lambda param1, param2, etc: expression

def multiply(x, y, z):
    return x * y * z

multiply(10,20,2)

400

In [14]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)

(lambda x,y,z: x * y * z)(2,3,4)

24

In [15]:
def accept_args(*args):
    return args
accept_args(1,2,3,4)

(1, 2, 3, 4)

In [17]:
(lambda *args: args)('a', 'b', 'c')

('a', 'b', 'c')

In [19]:
def accept_kwargs(**kwargs):
    return kwargs
accept_kwargs(name = "Brian", age = '99')

{'name': 'Brian', 'age': '99'}

In [20]:
(lambda **kwargs: kwargs)(test=123, name='Brian')

{'test': 123, 'name': 'Brian'}

In [ ]:
def get_evens(nums)

#### Passing a Function into a Function

In [21]:
def apply_to_each(f, a_list):
    """
    Print the return value of function f applied to every element in the list
    f -> Function to execute
    a_list -> List of elements to execute function f on
    """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [24]:
def get_first_letter(string):
    return string[0]

In [25]:
list_of_seasons = ['Winter', 'Spring', 'Summer', 'Fall']

In [26]:
apply_to_each(get_first_letter, list_of_seasons)

W
S
S
F


In [27]:
def reverse_string(string):
    return string[::-1]


apply_to_each(reverse_string, list_of_seasons)

retniW
gnirpS
remmuS
llaF


#### Passing a Lambda into a Function

In [32]:
# Call the apply_to_each function, pass in a lambda function as f argument

apply_to_each(lambda season: season.upper(), list_of_seasons)

WINTER
SPRING
SUMMER
FALL


In [33]:
apply_to_each([5, 10, 15, 20, 25, 30], lambda num: num**2 if num % 2 == 0 else num * 2)

TypeError: 'function' object is not iterable

In [35]:
def square_num(x):
    return x**2

apply_to_each(square_num, [1,2,3,4,5])

1
4
9
16
25


In [36]:
apply_to_each(lambda x: x**2, [1, 2, 3, 4, 5])

1
4
9
16
25


#### Returning a Function from a Function

In [46]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    # Define a locally scoped function
    def my_inner_function():
        print('This is the inner function!!!')
    # execute the inner function
    my_inner_function()
    
    
print(my_outer_function)
my_outer_function

<function my_outer_function at 0x00000213C768CB80>


<function __main__.my_outer_function(name)>

In [47]:
my_outer_function('Brian')

Hello Brian
This is the inner function!!!


In [51]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    # Define a locally scoped function
    def my_inner_function():
        print('This is the inner function!')
    
    return my_inner_function
    
    
print(my_outer_function)
my_outer_function

<function my_outer_function at 0x00000213C81F0360>


<function __main__.my_outer_function(name)>

In [52]:
my_outer_function_return = my_outer_function('Brian')
print(my_outer_function_return)

Hello Brian
<function my_outer_function.<locals>.my_inner_function at 0x00000213C81F04A0>


In [53]:
my_outer_function_return()

This is the inner function!


In [54]:
def create_multiplier(num):
    def multiply(x):
        return x * num
    return multiply


five_times = create_multiplier(5)
print(five_times)

<function create_multiplier.<locals>.multiply at 0x00000213C81F0040>


In [56]:
five_times(10)

50

In [58]:
five_times(3)

15

In [60]:
three_times = create_multiplier(3)
print(three_times)

<function create_multiplier.<locals>.multiply at 0x00000213C81F00E0>


In [62]:
three_times(10)

30

In [63]:
three_times(3)

9

#### Returning a Lambda from a Function

In [64]:
def new_multiplier(num):
    return lambda x: x * num


four = new_multiplier(4)
print(four)

<function new_multiplier.<locals>.<lambda> at 0x00000213C81F0180>


In [66]:
four(10)

40

In [67]:
four(5)

20

#### If Statements within Lambdas

In [68]:
def double_pos_square_negs(number):
    if number >= 0:
        return number * 2
    else:
        return number ** 2
    
    
print(double_pos_square_negs(10))
print(double_pos_square_negs(20))
print(double_pos_square_negs(-10))
print(double_pos_square_negs(-20))

20
40
100
400


In [69]:
# Ternary: value_if_true if condition else value_if_false

def double_pos_square_negs(number):
    #     if number >= 0:
    #         return number * 2
    #     else:
    #         return number ** 2
    return number * 2 if number >= 0 else number **2

    
    
print(double_pos_square_negs(10))
print(double_pos_square_negs(20))
print(double_pos_square_negs(-10))
print(double_pos_square_negs(-20))

20
40
100
400


In [70]:
(lambda num: num * 2 if num >= 0 else num ** 2)(10)

20

In [71]:
(lambda num: num * 2 if num >= 0 else num ** 2)(-10)

100

#### In-Class Exercise #1 <br>
<p>Write a lambda function that takes in three arguments <code>a</code>,<code>b</code>,<code>c</code> and returns the product of <code>a</code> and the difference of <code>b</code> and <code>c</code></p>

In [82]:
# Example 1 
# a = 10, b = 5, c = 2
# Output: 30

# Example 2
# a = 20, b = 10, c = 5
# Output: 100

func = lambda a,b,c: a * (b-c)
print(func(10, 5, 2))
print(func(20, 10, 5))

30
100


In [84]:
(lambda a,b,c: a*(b-c))(10,5,2)

30

In [88]:
(lambda a,b,c: a*(b-c))(20,10,5)

100

In [89]:
import re

In [90]:
# week4 day1 whiteboard solution alternative
def max_consecutives(nums):
    str_nums = ''.join([str(num) for num in nums])
    ones = re.compile('1+')
    one_groups = ones.findall(str_nums)
    return max([len(ones) for ones in one_groups])

print(max_consecutives([1,1,1,1,1,0,1,1,0,1]))
    

5


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [91]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [93]:
# Function that we want to apply to each element in an iterable
def squared_minus_two(num):
    return num ** 2 - 2

numbers = [3, 5, 7, 2, 4, 3, 7, 6]

# Create an empty list
new_nums = []
# Loop over the numbers list
for n in numbers:
    # Apply the function to that element
    new_num = squared_minus_two(n)
    new_nums.append(new_num)
    
print(new_nums)

[7, 23, 47, 2, 14, 7, 47, 34]


In [94]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)

list(map(squared_minus_two, numbers))

[7, 23, 47, 2, 14, 7, 47, 34]

In [95]:
my_map_obj = map(squared_minus_two, numbers)
print(my_map_obj)

In [96]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [106]:
next(my_map_obj) #every time I pressed run button it gave one index
# from the function above. First gave 7, then 23, then 47..., 34 and error

StopIteration: 

In [107]:
my_map_obj = map(squared_minus_two, numbers)

for num in my_map_obj:
    print(num)

7
23
47
2
14
7
47
34


In [108]:
x_values = [1, 2, 3, 4, 5, 6, 7]
y_values = [10, 20, 30, 40, 50, 60, 70]

def multiply(x, y):
    print(f"x={x}, y={y}")
    return x * y


list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
x=6, y=60
x=7, y=70


[10, 40, 90, 160, 250, 360, 490]

In [109]:
x_values = [1, 2, 3, 4, 5, 6, 7] # it only works until the shorter list is exhausted.
y_values = [10, 20, 30, 40, 50, 60, 70, 80, 90] #it doesn't matter which one is shorter

def multiply(x, y):
    print(f"x={x}, y={y}")
    return x * y


list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
x=6, y=60
x=7, y=70


[10, 40, 90, 160, 250, 360, 490]

In [110]:
nums_mult = []
shortest_length = len(x_values) if len(x_values) < len(y_values) else len(y_values)
for i in range(shortest_length):
    nums_mult.append(x_values[i] * y_values[i])
    
print(nums_mult)

[10, 40, 90, 160, 250, 360, 490]


In [111]:
def hello_func(element):
    return element.upper()
my_list = ['a', 'b', 'c']
list(map(hello_func, my_list))

['A', 'B', 'C']

#### Using Lambda's with Map

In [114]:
my_list2 = ['a', 'b', 'c']

upper_letters = list(map(lambda x: x.upper(), my_list2))
print(upper_letters)

['A', 'B', 'C']


In [115]:
x_vals = [1, 2, 3, 4, 5]
y_vals = [10, 20, 30, 40, 50]

list(map(lambda a,b: a*b, x_vals, y_vals))

[10, 40, 90, 160, 250]

In [116]:
random_letters = 'abcdefghijklmnopqrstuvwxyz'

unicode = list(map(lambda letter: ord(letter), random_letters))

print(unicode)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122]


In [118]:
vowels = ('a', 'e', 'i', 'o', 'u')

tuple(map(lambda letter: letter.upper(), vowels))

('A', 'E', 'I', 'O', 'U')

In [120]:
squares = list(map(lambda x: x**2, range(10,21)))
print(squares)

[100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if it is odd</p>

In [128]:
numbers = range(1,11)

# [0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]

x = list(map(lambda x: x*2 if x % 2 == 0 else x/2, numbers))
print(x)

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [129]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [130]:
names = ['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(map(starts_with_a, names))
print(a_names_map)
# map function takes the return boolean -> value True or False

[False, False, True, False, True, True, False, False]


In [131]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
# filter func -> if value is False it removes, if True stays!
a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['Alex', 'Amy', 'Andy']


In [132]:
def remainder(num):
    # Takes in a number and returns the remainder after dividing by 3
    return num % 3

my_nums = [x for x in range(1,11)]
print(my_nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [133]:
list(map(remainder, my_nums)) # prints the leftover of the division above

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]

In [134]:
list(filter(remainder, my_nums)) #removed the 3,6,9

[1, 2, 4, 5, 7, 8, 10]

In [135]:
for i in range(-10, 11):
    print(i, bool(i))

-10 True
-9 True
-8 True
-7 True
-6 True
-5 True
-4 True
-3 True
-2 True
-1 True
0 False
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True


#### Using Lambda's with Filter()

In [136]:
list(filter(lambda name: True if name[0].upper() == 'B' else False, names))

['Bob', 'Billy', 'Barbara']

In [137]:
list(filter(lambda name: name[0].upper() == 'C', names))

['Catherine', 'Cameron']

In [138]:
def get_c_names(names):
    return list(filter(lambda n: n[0].upper() == 'C', names))

get_c_names(['Cindy', 'Margaret', 'Michael', 'Chelsea', 'Jack', 'Connor', 'Colin'])

['Cindy', 'Chelsea', 'Connor', 'Colin']

#### In-Class Exercise #3 <br>
<p>Write a function that takes in a list of cities as strings and returns a list of cities with at least two words in the name</p>

In [139]:
import re

In [144]:
cities = ['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']
# Output: [Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge'] 

two_cities = list(filter(lambda string: re.search('\s', string), cities))
print(two_cities)


['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

In [145]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [146]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



#### Syntax

In [147]:
def add_nums(x, y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns {x + y}")
    return x + y

result_add = reduce(add_nums, [1, 2, 3, 4, 5])
print(result_add)

x: 1
y: 2
returns 3
x: 3
y: 3
returns 6
x: 6
y: 4
returns 10
x: 10
y: 5
returns 15
15


In [148]:
total = 0
for num in [1, 2, 3, 4, 5]:
    total += num
print(total)

15


In [149]:
def find_max(a,b):
    print('='*20)
    print(f"a: {a}")
    print(f'b: {b}')

#### Using Lambda's with Reduce()

In [150]:
lambda_reduce = reduce(lambda a,b: a+b, [10, 20, 30, 40, 50])
lambda_reduce

150

In [151]:
max_value = reduce(lambda x, y: x if x > y else y, [10, 5, 8, 12, 9, 4])
print(max_value)

12


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [158]:
nums = [2, 4, 6, 8, 10]
# Output: 3840

multiply = reduce(lambda x,y: x*y, nums)
print(multiply)

3840


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [ ]:
# Recursive function is a function that calls itself

In [159]:
def capitalize(name):
    # Takes in a name, upper case the first letter and lower case the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize the first name, take in a last name, capitalize the last name and 
    # then returns the concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)


def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
greet_person('bRIaN', 'stANtOn', 'tUEsdaY')

Hello Brian Stanton, it sure is beautiful on this Tuesday


In [160]:
import time

def func_one():
    print("Function One Started")
    time.sleep(5)
    print("Function One Ended")
    
    
def func_two():
    print("Function Two Started")
    func_one()
    print("Function Two Ended")
    
    
func_two()

Function Two Started
Function One Started
Function One Ended
Function Two Ended


#### Implementing a Base Case

In [161]:
def add_nums(num):
    # Implement a Base Case
    if num <= 1:
        return num
    else:
        # Add the num to return of the add_nums function with num minus 1
        return num + add_nums(num - 1)
    
    
add_nums(5)  

15

In [162]:
def add_nums(num):
    # Set up a base case
    if num <= 1:
        print(f"add_nums({num}) = {num}")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
    
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

In [166]:
def is_even(num):
    if num == 1:
        return False
    elif num == 0:
        return True
    else:
        return is_even(num - 2)
# is_even(11)
# is_even(10)
# is_even(22)
is_even(23)

False

In [168]:
def print_recursion(a_list):
    if a_list:
        print(a_list[0])
        print_recursion(a_list[1:])
    
    
print_recursion([10, 20, 30, 40, 50])

10
20
30
40
50


#### Writing a Factorial Function

In [169]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

# factorial(x) = x * factorial(x-1)

def factorial(num):
    if num <= 1: # BASE CASE
#         print(f"factorial({num}) = 1")
        return 1 #BASE CASE
    else:
#         print(f"factorial({num}) = {num} * factorial({num - 1})")
        return num * factorial(num - 1)
    

factorial(5)     

120

In [ ]:
# NEEDS A BASE CASE otherwise doesn't work!

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [170]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple'])

print(lengths)

In [171]:
for item in lengths:
    print(item)

5
7
13


In [172]:
for item in lengths:
    print(item)
    
    #PRINTS NOTHING because that's how Generators work (map???)
    # returns yeld because it doesn't store the value, just the "result of what len is calling"

In [173]:
help(iter)

Help on built-in function iter in module builtins:

iter(...)
    iter(iterable) -> iterator
    iter(callable, sentinel) -> iterator
    
    Get an iterator from an object.  In the first form, the argument must
    supply its own iterator, or be a sequence.
    In the second form, the callable is called until it returns the sentinel.



In [174]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [175]:
a_new_list = ['hi', 'hello', 'how are you', 'goodbye']

list_iter = iter(a_new_list)
print(list_iter)

In [176]:
next(list_iter)

'hi'

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [179]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        return start
        start += step
        
my_range_iter(0, 10)

0

In [180]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        yield start
        start += step
        
my_range_iter(0, 10)

<generator object my_range_iter at 0x00000213C830AB50>

In [182]:
my_gen = my_range_iter(0,10)
print(my_gen)

<generator object my_range_iter at 0x00000213C83086C0>


In [183]:
next(my_gen)

0

In [184]:
for s in my_range_iter(10, 50, 5):
    print(s)

10
15
20
25
30
35
40
45


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC"]


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [193]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]
# case insensitive

pattern = re.compile((?P<first>[A-z]{2,})' '(?P<last>[A-z]{2,})')


[]


In [ ]:
cities = ['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']
# Output: [Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge'] 

two_cities = list(filter(lambda string: re.search('\s', string), cities))
print(two_cities)

In [185]:
# use .sort() or sorted()
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [186]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]

def capitalize(name):
    # Takes in a name, upper case the first letter and lower case the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize the first name, take in a last name, capitalize the last name and 
    # then returns the concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)


def last_name(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
greet_person('bRIaN', 'stANtOn', 'tUEsdaY')

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

```
Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
```

In [ ]:
def fib(num):
    pass

In [ ]:
n = 6

for i in range(n):
    print(f"Iteration {i}: {fib(i)}")

### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared, then call the generator and iterate through and print the result.</p>

In [ ]:

# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0